src/u2f.c

```diff
@@ -11,477 +11,606 @@
+static void _cmd_sync(const Packet* in_packet, Packet* out_packet, const size_t max_out_len)
+{
+    (void)in_packet;
+    (void)out_packet;
+    (void)max_out_len;
+}
+
+static void _cmd_lock(const Packet* in_packet, Packet* out_packet, const size_t max_out_len)
+{
+    (void)in_packet;
+    (void)out_packet;
+    (void)max_out_len;
+}
+*/
+
+/**
+ * Synchronize a channel and optionally requests a unique 32-bit channel identifier (CID)
+ * that can be used by the requesting application during its lifetime.
+ *
+ * If the CID is the U2FHID_CID_BROADCAST then the application is requesting a CID from the device.
+ * Otherwise the application has chosen the CID.
+ */
+static void _cmd_init(const Packet* in_packet, Packet* out_packet, const size_t max_out_len)
+{
+    if (U2FHID_INIT_RESP_SIZE >= max_out_len) {
+        _error_hid(in_packet->cid, U2FHID_ERR_OTHER, out_packet);
+        return;
+    }
+
+    // Channel 0 is reserved
+    if (in_packet->cid == 0) {
+        _error_hid(in_packet->cid, U2FHID_ERR_INVALID_CID, out_packet);
+        return;
+    }
+
+    const U2FHID_INIT_REQ* init_req = (const U2FHID_INIT_REQ*)&in_packet->data_addr;
+    U2FHID_INIT_RESP response;
+
+    out_packet->cid = in_packet->cid;
+    out_packet->cmd = U2FHID_INIT;
+    out_packet->len = U2FHID_INIT_RESP_SIZE;
+
+    util_zero(&response, sizeof(response));
+    memcpy(response.nonce, init_req->nonce, sizeof(init_req->nonce));
+    response.cid = in_packet->cid == U2FHID_CID_BROADCAST ? _next_cid() : in_packet->cid;
+    response.versionInterface = U2FHID_IF_VERSION;
+    response.versionMajor = DIGITAL_BITBOX_VERSION_MAJOR;
+    response.versionMinor = DIGITAL_BITBOX_VERSION_MINOR;
+    response.versionBuild = DIGITAL_BITBOX_VERSION_PATCH;
+    response.capFlags = U2FHID_CAPFLAG_WINK;
+    util_zero(out_packet->data_addr, sizeof(out_packet->data_addr));
+    memcpy(out_packet->data_addr, &response, sizeof(response));
+}
+
+/**
+ * Process a U2F message
+ */
+// TODO - likely need to handle and return appropriate error messages for U2F specification
+static void _cmd_msg(const Packet* in_packet, Packet* out_packet, const size_t max_out_len)
+{
+    (void)max_out_len;
+    // By default always use the recieved cid
+    _cid = in_packet->cid;
+
+    const USB_APDU* apdu = (const USB_APDU*)in_packet->data_addr;
+
+    if ((APDU_LEN(*apdu) + sizeof(USB_APDU)) > in_packet->len) {
+        return;
+    }
+
+    if (apdu->cla != 0) {
+        _error(U2F_SW_CLA_NOT_SUPPORTED, out_packet);
+        return;
+    }
+
+    switch (apdu->ins) {
+    case U2F_REGISTER:
+        _register(apdu, out_packet);
+        break;
+    case U2F_AUTHENTICATE:
+        _authenticate(apdu, out_packet);
+        break;
+    case U2F_VERSION:
+        _version(apdu, out_packet);
+        break;
+    default:
+        _error(U2F_SW_INS_NOT_SUPPORTED, out_packet);
+        return;
+    }
+}
+
+// TODO - currently not called - see v1 code for example usage
```

Comment:
delete? seems like you solved timeouts differently

Reply:
done